microservice/package/src/main/java/com/cumulocity/agent/packaging/microservice/impl/MicroserviceDockerClientImpl.java

```diff
@@ -1,81 +1,258 @@
+        }
+
+        if (!Strings.isNullOrEmpty(networkMode)) {
+            buildImageCmd = buildImageCmd.withNetworkMode(networkMode);
+        }
+
+        log.info("Building Docker image. Docker dir={},tags={}, build arguments={}", dockerDirectory, tags, buildArgs);
+        ImageBuildCompletionWaiter imageBuildCompletionWaiter = new ImageBuildCompletionWaiter();
+        buildImageCmd.exec(imageBuildCompletionWaiter);
+        log.info("Waiting for image build to complete.");
+        imageBuildCompletionWaiter.waitUntilCompletionOrFailure();
+
+        if (Objects.nonNull(imageBuildCompletionWaiter.getImageBuildError())) {
+            log.error("Image build failed", imageBuildCompletionWaiter.getImageBuildError());
+            throw imageBuildCompletionWaiter.getImageBuildError();
+        } else {
+            log.info("Image build success");
+        }
+
+    }
+
+
+    protected class ImageBuildCompletionWaiter extends BuildImageResultCallback {
+
+
+        //As the image build is asynchronous, we use active waiting for simplicity reasons
+        @Getter
+        private AtomicBoolean imageBuildOperationCompleted = new AtomicBoolean(false);
+
+        @SneakyThrows
+        public void waitUntilCompletionOrFailure() {
+            while (imageBuildOperationCompleted.get()==false) {
+                Thread.sleep(250);
+            }
+        }
+
+        @Getter
+        private Throwable imageBuildError = null;
+
+        //this logs out the build progress
+        @Override
+        public void onNext(BuildResponseItem item) {
+            if (Objects.nonNull(item.getStream())){
+                log.info(item.getStream());
+            }
+        }
+
+        @Override
+        public void onError(Throwable throwable) {
+            imageBuildError=throwable;
+            imageBuildOperationCompleted.set(true);
+        }
+
+        @Override
+        public void onComplete() {
+            imageBuildOperationCompleted.set(true);
+        }
+    }
+
+    public void saveDockerImage(final String image, final File targetFile) throws IOException {
+
+        log.info("Saving Image {} to file {}", image, targetFile.getAbsoluteFile().toString());
+
         try (
-                final InputStream inputStream = docker.save(image);
+                SaveImageCmd saveImageCmd = dockerClient.saveImageCmd(image);
+
+                final InputStream inputStream = saveImageCmd.exec();
                 final FileOutputStream outputStream = new FileOutputStream(targetFile)
         ) {
             IOUtils.copyLarge(inputStream, outputStream);
         }
     }
 
-    private DefaultDockerClient newDockerClient() throws DockerCertificateException {
-        return DefaultDockerClient.fromEnv().build();
+    private static DockerClient newDockerClient() {
+
+        DockerClientConfig config = DefaultDockerClientConfig.createDefaultConfigBuilder().build();
+
+        System.out.println(config.toString());
```

Comment:
sysout?

Reply:
No, debugging leftover :D I  will remove that.

Reply:
removed with lastest commit